common/rpc/interceptor/namespace_validator.go

```diff
@@ -0,0 +1,142 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interceptor
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/grpc"
+
+	"go.temporal.io/server/common/namespace"
+)
+
+type (
+	// NamespaceValidatorInterceptor validates:
+	// 1. Namespace is specified in request if there is a `namespace` field.
+	// 2. Namespace exists.
+	// 3. Namespace is in correct state.
+	NamespaceValidatorInterceptor struct {
+		namespaceRegistry namespace.Registry
+	}
+)
+
+var (
+	ErrNamespaceNotSet              = serviceerror.NewInvalidArgument("Namespace not set on request.")
+	errInvalidNamespaceStateMessage = "Namespace has invalid state: %s. Must be %s."
+
+	// By default, enumspb.NAMESPACE_STATE_REGISTERED and enumspb.NAMESPACE_STATE_DEPRECATED are allowed states for all APIs that have `namespace` field in the request object.
+	allowedNamespaceStates = map[string][]enumspb.NamespaceState{
+		"StartWorkflowExecution":           {enumspb.NAMESPACE_STATE_REGISTERED},
+		"SignalWithStartWorkflowExecution": {enumspb.NAMESPACE_STATE_REGISTERED},
+		"DescribeNamespace":                {enumspb.NAMESPACE_STATE_REGISTERED, enumspb.NAMESPACE_STATE_DEPRECATED, enumspb.NAMESPACE_STATE_DELETED},
```

Comment:
Allowed states per API.

Reply:
I don't think we should support NAMESPACE_STATE_DEPRECATED now. It is not used anywhere. It can be added here in future when DEPRECATED is supported.

Reply:
`DeprecateNamespace` API is there. And although I can tell for sure that no one ever used it (it panics), it is easier to fix and support it rather then removing.